File: packages/python/jobping/envelope.py
"""Transport-neutral JobPing result envelope."""

from __future__ import annotations

from typing import Any, Literal, TypedDict


JOBPING_ENVELOPE_KIND = "jobping.envelope.v1"
JOBPING_RESULT = "result"

EnvelopeType = Literal["result"]


class ResultEnvelope(TypedDict):
    jobping: str
    type: Literal["result"]
    job_id: str
    payload: Any


JobPingEnvelope = ResultEnvelope


def _assert_valid_job_id(job_id: str) -> None:
    if not isinstance(job_id, str) or not job_id:
        raise ValueError("job_id must be a non-empty string")


def box_result(job_id: str, payload: Any) -> ResultEnvelope:
    _assert_valid_job_id(job_id)

    return {
        "jobping": JOBPING_ENVELOPE_KIND,
        "type": JOBPING_RESULT,
        "job_id": job_id,
        "payload": payload,
    }


def is_envelope(value: Any) -> bool:
    return (
        isinstance(value, dict)
        and value.get("jobping") == JOBPING_ENVELOPE_KIND
        and isinstance(value.get("type"), str)
        and isinstance(value.get("job_id"), str)
        and len(value["job_id"]) > 0
    )


def is_result_envelope(value: Any) -> bool:
    return is_envelope(value) and value["type"] == JOBPING_RESULT and "payload" in value


def unbox_result(envelope: Any, expected_job_id: str | None = None) -> Any:
    if not is_result_envelope(envelope):
        raise ValueError("Expected JobPing result envelope")

    if expected_job_id is not None and envelope["job_id"] != expected_job_id:
        raise ValueError("Unexpected JobPing result job_id")

    return envelope["payload"]

File: packages/js/envelope.mjs
// Transport-neutral JobPing result envelope.

export const JOBPING_ENVELOPE_KIND = "jobping.envelope.v1";
export const JOBPING_RESULT = "result";

function assertValidJobId(jobId) {
  if (typeof jobId !== "string" || jobId.length === 0) {
    throw new Error("job_id must be a non-empty string");
  }
}

export function boxResult(jobId, payload) {
  assertValidJobId(jobId);

  return {
    jobping: JOBPING_ENVELOPE_KIND,
    type: JOBPING_RESULT,
    job_id: jobId,
    payload,
  };
}

export function isEnvelope(value) {
  return (
    typeof value === "object" &&
    value !== null &&
    value.jobping === JOBPING_ENVELOPE_KIND &&
    typeof value.type === "string" &&
    typeof value.job_id === "string" &&
    value.job_id.length > 0
  );
}

export function isResultEnvelope(value) {
  return isEnvelope(value) && value.type === JOBPING_RESULT && Object.hasOwn(value, "payload");
}

export function unboxResult(envelope, expectedJobId) {
  if (!isResultEnvelope(envelope)) {
    throw new Error("Expected JobPing result envelope");
  }

  if (expectedJobId !== undefined && envelope.job_id !== expectedJobId) {
    throw new Error("Unexpected JobPing result job_id");
  }

  return envelope.payload;
}
